### PR TITLE
ConvexYieldWrapper: modified earned function to the latest one

### PR DESCRIPTION
**Pull request by iamsahu**
Fix for the issue:
- https://github.com/code-423n4/2022-01-yield-findings/issues/95

### CODE REVIEW

File: contracts/utils/convex/ConvexStakingWrapper.sol [hunk 1]
```diff
@@ -100,23 +99,40 @@ contract ConvexStakingWrapper is ERC20 {
...
-        uint256 rewardsLength = rewards.length;
-
-        if (rewardsLength == 0) {
+        if (rewards.length == 0) {
             RewardType storage reward = rewards.push();
             reward.reward_token = crv;
             reward.reward_pool = mainPool;
-            rewardsLength = 1;
+            reward.reward_integral = 0;
+            reward.reward_remaining = 0;
```
Comment (alcueca) [inline contracts/utils/convex/ConvexStakingWrapper.sol:107]: The two lines above can be removed.

File: contracts/utils/convex/ConvexStakingWrapper.sol [hunk 2]
```diff
@@ -100,23 +99,40 @@ contract ConvexStakingWrapper is ERC20 {
...
+
+            reward = rewards.push();
+            reward.reward_token = cvx;
+            reward.reward_pool = address(0); // This is set to address(0) as initially we don't know if the pool has cvx rewards
```
Comment (alcueca) [inline contracts/utils/convex/ConvexStakingWrapper.sol:111]: This line can be replaced by just the comment.

File: contracts/utils/convex/ConvexStakingWrapper.sol [hunk 3]
```diff
@@ -100,23 +99,40 @@ contract ConvexStakingWrapper is ERC20 {
...
+            reward.reward_integral = 0;
+            reward.reward_remaining = 0;
```
Comment (alcueca) [inline contracts/utils/convex/ConvexStakingWrapper.sol:113]: The two lines above can be removed.

File: contracts/utils/convex/ConvexStakingWrapper.sol [hunk 4]
```diff
@@ -100,23 +99,40 @@ contract ConvexStakingWrapper is ERC20 {
...
+
+            registeredRewards[crv] = CRV_INDEX + 1; //mark registered at index+1
+            registeredRewards[cvx] = CVX_INDEX + 1; //mark registered at index+1
```
Comment (alcueca) [inline contracts/utils/convex/ConvexStakingWrapper.sol:116]: Why +1? Can't you register them at the index, and then subtract 1 to rewards.length when adding marks for other tokens?

File: contracts/utils/convex/ConvexStakingWrapper.sol [hunk 5]
```diff
@@ -100,23 +99,40 @@ contract ConvexStakingWrapper is ERC20 {
...
         }
 
         uint256 extraCount = IRewardStaking(mainPool).extraRewardsLength();
-
-        uint256 startIndex = rewardsLength - 1;
-        for (uint256 i = startIndex; i < extraCount; i++) {
+        for (uint256 i = 0; i < extraCount; i++) {
```
Comment (alcueca) [inline contracts/utils/convex/ConvexStakingWrapper.sol:120]: Can we remove the use of `startIndex`?

File: contracts/utils/convex/ConvexStakingWrapper.sol [hunk 6]
```diff
@@ -100,23 +99,40 @@ contract ConvexStakingWrapper is ERC20 {
...
             address extraPool = IRewardStaking(mainPool).extraRewards(i);
-            RewardType storage reward = rewards.push();
-            reward.reward_token = IRewardStaking(extraPool).rewardToken();
-            reward.reward_pool = extraPool;
+            address extraToken = IRewardStaking(extraPool).rewardToken();
+            if (extraToken == cvx) {
+                //update cvx reward pool address
+                rewards[CVX_INDEX].reward_pool = extraPool;
+            } else if (registeredRewards[extraToken] == 0) {
```
Comment (alcueca) [inline contracts/utils/convex/ConvexStakingWrapper.sol:126]: This means that the CVX pool can be updated, but not the pool for other tokens. That is inconsistent.

File: contracts/utils/convex/ConvexStakingWrapper.sol [hunk 7]
```diff
@@ -100,23 +99,40 @@ contract ConvexStakingWrapper is ERC20 {
...
+                //add new token to list
+                RewardType storage reward = rewards.push();
+                reward.reward_token = extraToken;
+                reward.reward_pool = extraPool;
+                reward.reward_integral = 0;
+                reward.reward_remaining = 0;
```
Comment (alcueca) [inline contracts/utils/convex/ConvexStakingWrapper.sol:132]: The two lines above can be removed.

File: contracts/utils/convex/ConvexStakingWrapper.sol [hunk 8]
```diff
@@ -51,6 +49,7 @@ contract ConvexStakingWrapper is ERC20 {
 
     //rewards
     RewardType[] public rewards;
+    mapping(address => uint256) public registeredRewards;
```
Comment (alcueca) [inline contracts/utils/convex/ConvexStakingWrapper.sol:52]: What's this for?

File: contracts/utils/convex/ConvexStakingWrapper.sol [hunk 9]
```diff
@@ -308,15 +271,23 @@ contract ConvexStakingWrapper is ERC20 {
     /// @return claimable Array of earned tokens and their amount
     function earned(address _account) external view returns (EarnedData[] memory claimable) {
         uint256 supply = _getTotalSupply();
+        // uint256 depositedBalance = _getDepositedBalance(_account);
         uint256 rewardCount = rewards.length;
-        claimable = new EarnedData[](rewardCount + 1);
+        claimable = new EarnedData[](rewardCount);
```
Comment (alcueca) [inline contracts/utils/convex/ConvexStakingWrapper.sol:276]: Mmm, why?

File: contracts/utils/convex/ConvexStakingWrapper.sol [hunk 10]
```diff
@@ -308,15 +271,23 @@ contract ConvexStakingWrapper is ERC20 {
...
 
-        for (uint256 i; i < rewardCount; ++i) {
+        for (uint256 i = 0; i < rewardCount; i++) {
```
Comment (alcueca) [inline contracts/utils/convex/ConvexStakingWrapper.sol:278]: No need to initialize `i` if it is zero.

In my implementation I initialized it to `1`, and did the `cvx` block before the loop. Now I see that I might be missing something.

File: contracts/utils/convex/ConvexStakingWrapper.sol [hunk 11]
```diff
@@ -308,15 +271,23 @@ contract ConvexStakingWrapper is ERC20 {
...
             RewardType storage reward = rewards[i];
-            address rewardToken = reward.reward_token;
+
+            if (reward.reward_pool == address(0)) {
+                //cvx reward may not have a reward pool yet
+                //so just add whats already been checkpointed
```
Comment (alcueca) [inline contracts/utils/convex/ConvexStakingWrapper.sol:283]: Don't we add what's already been checkpointed if cvx has a reward pool?

**Review by alcueca — CHANGES_REQUESTED** [review]
Very interesting refactor. I've got lots of questions before I understand all of it, though.

File: contracts/utils/convex/ConvexStakingWrapper.sol (hunk 10 above)
Comment (iamsahu) [reply contracts/utils/convex/ConvexStakingWrapper.sol:278]: If we do this then if there is any new claimable reward that is there for `cvx` would be missed. The one calculated [here](https://github.com/yieldprotocol/vault-v2/blob/45256b8aca37b23b306b2a09666a5e324e14e0a9/contracts/utils/convex/ConvexStakingWrapper.sol#L299)

File: contracts/utils/convex/ConvexStakingWrapper.sol (hunk 8 above)
Comment (iamsahu) [reply contracts/utils/convex/ConvexStakingWrapper.sol:52]: Used to keep track of the rewards that are added. Makes the `addRewards` robust & prevent any duplication.

File: contracts/utils/convex/ConvexStakingWrapper.sol [hunk 12]
```diff
@@ -100,23 +99,35 @@ contract ConvexStakingWrapper is ERC20 {
...
-        for (uint256 i = startIndex; i < extraCount; i++) {
+        for (uint256 i; i < extraCount; ++i) {
             address extraPool = IRewardStaking(mainPool).extraRewards(i);
-            RewardType storage reward = rewards.push();
-            reward.reward_token = IRewardStaking(extraPool).rewardToken();
-            reward.reward_pool = extraPool;
+            address extraToken = IRewardStaking(extraPool).rewardToken();
+            if (extraToken == cvx) {
+                //update cvx reward pool address
+                rewards[CVX_INDEX].reward_pool = extraPool;
```
Comment (alcueca) [inline contracts/utils/convex/ConvexStakingWrapper.sol:122]:
```suggestion
                if (rewards[CVX_INDEX].reward_pool == address(0)) {
                    rewards[CVX_INDEX].reward_pool = extraPool;
                }
```

**Review by alcueca — CHANGES_REQUESTED** [review]
Let's ship 🚢

**Review by alcueca — APPROVED** [review]
Oops, wrong button

**Pull request merged**